### PR TITLE
Changes to validation

### PR DESCRIPTION
**Pull request by Jurigag**
What i changed in this commit:
- Added option for adding multiple fields to validate/filters with one(same) validator/filter - also it checking for messages, labels etc, so if we have 10 `Phalcon\Validation\Validator\PresenceOf` validators - we can shorten it to one add method and use array of fields, messages etc, it's working for every option except `model` in `Phalcon\Validation\Validator\Uniqueness` validator(if i missed something then write it here)
- In `Phalcon\Validation\Validator\Uniqueness::isUniqueness()` i added getting Entity from validation if we don't provide any `model` so we can get rid of "unlogical" `model => $this` when we validating from model
- Changed checking for strict in `Phalcon\Validation\Validator\ExclusionIn` and `Phalcon\Validation\Validator\InclusionIn` - it was checking for boolean value before getting it, i think it was wrong behaviour

What needs to be done:
- [x] tests
- [x] update documentation in phpdoc (examples)
- [ ] performance tests ?

### CODE REVIEW

**Comment by Besedin86** [comment]
I can't see in this PR how can I check combination of fields for uniqueness. It would be great if it will be possible.

**Comment by Jurigag** [comment]
Just by:

```php
$validation->add(['field1','field2'], new Uniqueness);
```

**Comment by Besedin86** [comment]
f you see [here](https://github.com/phalcon/cphalcon/pull/11826/files#diff-86920296fe08950334f3b3f6d672f12cR170), you just add validator for each field separately. But I ask about combination of fields: when `field1 = value1 and field2 = value2`, not `field1 = value1 or field2 = value2`. It think this is different things.

**Comment by Jurigag** [comment]
But it's not OR, it's still and, validation i succesfull when all validatiors are successfull.

**Comment by Besedin86** [comment]
See example:

```sql
CREATE TABLE IF NOT EXISTS `test_table` (
  `id` int(11) NOT NULL,
  `field_1` int(11) NOT NULL,
  `field_2` int(11) NOT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=latin1;

INSERT INTO `test_table` (`id`, `field_1`, `field_2`) VALUES
	(1, 1, 1),
	(2, 1, 2),
	(3, 2, 1);
```

I want to check for uniqueness combination `field_1` and `field_2`. For example I want to add values 2 and 2.

```sql
INSERT INTO `test_table` (`id`, `field_1`, `field_2`) VALUES (4, 2, 2);
```

First validator will fails (I have 2 in record no. 3) and second validator will fails too (I have 2 in record no.2). But I have no record with values 2 2 simultaneously. Therefore you need build 
```sql
SELECT COUNT(*) FROM test_table WHERE field_1 = 2 AND field_2 = 2.
```

**Comment by Jurigag** [comment]
Yea after some thinking i think i got it. Will do commit soon with fix for it.

**Comment by Jurigag** [comment]
How except should work for combination of field ? It should be `OR` or `AND` ?

**Comment by Besedin86** [comment]
As I wrote - AND.

**Comment by Jurigag** [comment]
I mean this EXCEPT OPTION https://github.com/phalcon/cphalcon/blob/2.1.x/phalcon/validation/validator/uniqueness.zep#L106

For both cases AND ?

**Comment by Jurigag** [comment]
@sergeyklay 

Im not sure but shouldn't here:

https://github.com/phalcon/cphalcon/blob/2.1.x/phalcon/validation/validator/uniqueness.zep#L102

Be: 
```
if !except {

}
else {

}
```

**Comment by Besedin86** [comment]
I work around query 
```sql
select * from test_table 
where field_1= 2 and field_2 = 2 and field_1 <> 2 and field_2 <> 2
```
and it works as we expect.

**Comment by Jurigag** [comment]
Okay. Just wasn't sure about logic. What about above ? Not sure if this except even works as it should be ? Anyone even tested it ?

**Comment by Besedin86** [comment]
I've never used it. But I'm sure it will work correctly. One small moment: if we use set of fields - except must be array too.

**Comment by Jurigag** [comment]
Oh so this except just check additionally against our values. Roger that.

**Comment by Besedin86** [comment]
Also you have to check type of `except` param.

**Comment by Jurigag** [comment]
Oh right, except can be just string too.

**Comment by Besedin86** [comment]
If `except` is array, I think it preferred be indexed array to avoid mistakes.

**Comment by Jurigag** [comment]
So basically only this:

```php
'except'=>[
'field1'=>['value1'],
'field2'=>['value2']
]
```

Yea, i think the same.

**Comment by Besedin86** [comment]
I fixed my query. So final version is
```sql
select * from test_table 
where field_1= 2 and field_2 = 2 and # field values
      (field_1 <> 2 or field_2 <> 2) # except
```

This is right query.

**Comment by Jurigag** [comment]
Yea i know it's right. But what IF someone want fields not to be array of values ? I think it's logically possible, current uniqueness validator don't allow such a case.

**Comment by Besedin86** [comment]
The query will be the same

```sql
select * from test_table 
where (field_1= 2) and   # field values
      (field_1 <> 2)     # except
```

**Comment by Jurigag** [comment]
You don't understand, i mean such a case:

```php
$validation->add('field1', new Uniqueness([
'except'=>['value1','value2']
]));
```

Now it doesn't works. We can only check against one value(why ?).

**Comment by Besedin86** [comment]
Oh. So we can modify our script to generate such sql query
```sql
select count(*) from test_table 
where (field_1 = 2 and field_2 = 2) and # field values
      ((field_1 <> 1 and field_1 <> 3) or (field_2 <> 2)) # except
```
it exclude 1,2 and 3,2 combination of values.

**Comment by Jurigag** [comment]
Yea exactly i mean this. So basically i will add abiltiy to except value for field can be array of value.(now it can be only ONE VALUE)

**Comment by Besedin86** [comment]
So that we have such difference of formats:

```php
// single field, except - single value
$validation->add('field1', new Uniqueness([
'except'=> 'value1'
]));

// single field, except - array of values
$validation->add('field1', new Uniqueness([
'except'=> ['value1', 'value2']
]));

// set of fields, except - single value
$validation->add(['field1', 'field2'], new Uniqueness([
'except'=> ['field1' => 'value1']
]));

// set of fields, except - array of values
$validation->add(['field1', 'field2'], new Uniqueness([
'except'=> ['field1' => ['value1', 'value2']]
]));

// set of fields, except - mixed values
$validation->add(['field1', 'field2'], new Uniqueness([
'except'=> ['field1' => ['value1', 'value2'], 'field2' => 'value3']
]));

```

**Comment by Besedin86** [comment]
Optimized variant of query
```sql
select count(*) from test_table 
where (field_1 = 2 and field_2 = 2) and # field values
      ((field_1 not in (1,3)) or (field_2 not in (2))) # except
```

**Comment by Jurigag** [comment]
I think really it should be not in AND not in ? So basically there are always be ands ? Well from mysql point there is no difference in performance between NOT IN and <> i think.

**Comment by Besedin86** [comment]
It will be harder to parse query with `and and and and....`

**Comment by Jurigag** [comment]
Yea you right. I just didn't understand it correctly. Will do commit soon after quick testing of just uniqueness.

**Comment by Jurigag** [comment]
Almost done, but what Phalcon\Mvc\Model\Message field property should contain if we have two fields ? Just array or joined fields ?

**Comment by Besedin86** [comment]
I think joined fields with general message

**Comment by Jurigag** [comment]
Well as i see in old validation it was array of fields.

**Comment by Jurigag** [comment]
@Besedin86 

now uniqueness validator should as it should be

**Comment by sergeyklay** [comment]
@Jurigag

**Comment by Jurigag** [comment]
Well, i will try to finish it today and tomorrow.

**Comment by Jurigag** [comment]
There are some validators with really many options. I hope i don't have to provide test for each case(each option) beacause it's just similar code ?

**Comment by sergeyklay** [comment]
do as many as you can

**Comment by Jurigag** [comment]
Can i change everywhere in validator docs array to [] ?

**Comment by sergeyklay** [comment]
yes

**Comment by Jurigag** [comment]
Done, only some performance tests to do to check if it affects performance(i doubt it).

File: phalcon/validation.zep [hunk 1]
```diff
@@ -162,30 +207,63 @@ class Validation extends Injectable implements ValidationInterface
...
 	{
-		let this->_validators[] = [field, validator];
+		var f;
+		if typeof field == "array" {
+			// Uniqueness validator for combination of fields is handled differently
+			if validator instanceof Uniqueness {
+				let this->_uniquenessValidatorsCombinedFields[] = [field, validator];
+			}
+			else {
+				for f in field {
```
Comment (andresgutierrez) [inline phalcon/validation.zep]: Could you please rename variable 'f' to 'singleField'?

File: phalcon/validation.zep [hunk 2]
```diff
@@ -478,13 +570,26 @@ class Validation extends Injectable implements ValidationInterface
 	/**
 	 * Internal validations, if it returns true, then skip the current validator
 	 */
-	protected function preChecking(string field, <ValidatorInterface> validator) -> boolean
+	protected function preChecking(var field, <ValidatorInterface> validator) -> boolean
 	{
-		if validator->getOption("allowEmpty", false) {
-			if method_exists(validator, "isAllowEmpty") {
-				return validator->isAllowEmpty(this, field);
+		var f;
```
Comment (andresgutierrez) [inline phalcon/validation.zep]: Same here

Comment (Jurigag) [inline phalcon/validation.zep]: Changed this f and e too everywhere they were(i hope).

File: tests/unit/Validation/Validator/BetweenTest.php [hunk 3]
```diff
@@ -0,0 +1,92 @@
+<?php
+/**
+ * Created by PhpStorm.
+ * User: Wojtek
+ * Date: 2016-06-05
+ * Time: 15:42
+ */
```
Comment (sergeyklay) [inline tests/unit/Validation/Validator/BetweenTest.php]: @Jurigag Could you please add standard header here?

File: tests/unit/Validation/Validator/AlphaTest.php [hunk 4]
```diff
@@ -0,0 +1,77 @@
+<?php
+/**
+ * Created by PhpStorm.
+ * User: Wojtek
+ * Date: 2016-06-05
+ * Time: 15:35
+ */
```
Comment (sergeyklay) [inline tests/unit/Validation/Validator/AlphaTest.php]: Same here

File: tests/unit/Validation/Validator/AlnumTest.php [hunk 5]
```diff
@@ -0,0 +1,81 @@
+<?php
+/**
+ * Created by PhpStorm.
+ * User: Wojtek
+ * Date: 2016-06-05
+ * Time: 15:21
+ */
```
Comment (sergeyklay) [inline tests/unit/Validation/Validator/AlnumTest.php]: Same here

File: tests/unit/Validation/Validator/ConfirmationTest.php [hunk 6]
```diff
@@ -0,0 +1,87 @@
+<?php
+/**
+ * Created by PhpStorm.
+ * User: Wojtek
+ * Date: 2016-06-05
+ * Time: 16:23
+ */
```
Comment (sergeyklay) [inline tests/unit/Validation/Validator/ConfirmationTest.php]: Same here

File: tests/unit/Validation/Validator/CreditCardTest.php [hunk 7]
```diff
@@ -0,0 +1,81 @@
+<?php
+/**
+ * Created by PhpStorm.
+ * User: Wojtek
+ * Date: 2016-06-05
+ * Time: 17:47
+ */
```
Comment (sergeyklay) [inline tests/unit/Validation/Validator/CreditCardTest.php]: Same here

File: tests/unit/Validation/Validator/DateTest.php [hunk 8]
```diff
@@ -0,0 +1,85 @@
+<?php
+/**
+ * Created by PhpStorm.
+ * User: Wojtek
+ * Date: 2016-06-05
+ * Time: 17:58
+ */
```
Comment (sergeyklay) [inline tests/unit/Validation/Validator/DateTest.php]: Same here

File: tests/unit/Validation/Validator/DigitTest.php [hunk 9]
```diff
@@ -0,0 +1,81 @@
+<?php
+/**
+ * Created by PhpStorm.
+ * User: Wojtek
+ * Date: 2016-06-05
+ * Time: 18:08
+ */
```
Comment (sergeyklay) [inline tests/unit/Validation/Validator/DigitTest.php]: Same here

File: tests/unit/Validation/Validator/EmailTest.php [hunk 10]
```diff
@@ -0,0 +1,81 @@
+<?php
+/**
+ * Created by PhpStorm.
+ * User: Wojtek
+ * Date: 2016-06-05
+ * Time: 18:15
+ */
```
Comment (sergeyklay) [inline tests/unit/Validation/Validator/EmailTest.php]: Same here

**Comment by sergeyklay** [comment]
@Jurigag Could you please rebase

**Comment by Jurigag** [comment]
Heh, i totally forgot about those phpstorm comments. Done.

**Comment by sergeyklay** [comment]
Thanks

**Pull request merged**